### DISCUSSION TITLE
G6 On new sensor start, data with timestamp before new sensor was started is sent to NS and AAPS

### DISCUSSION THREAD

**Discussion by andyrozman** [Navid200/xDrip#148] [Dexcom]
So the problem AAPS (3.0) has is, that when new G6 sensor is started, old data (data accumulated before new sensor was started) is being added to NS and send over broadcast. This then in turns stops Loop for 1.5 days (BgQualityPlugin), because we get duplicated data. 

This seems like similar issue to #1778 or at least solution might be the same (but that ticket is now low priority). Once new sensor is started, all data from before should be ignored and dropped.

**Comment by Navid200** [comment]
Is there a thread somewhere where this is discussed?
For example, an AAPS discussion somewhere?  If yes, would you please add the link here?
Just to have everything related in one place before I point the developers to this.

**Reply by andyrozman** [reply]
Not really. I had to get details directly from main developer (Milos). I am just one of driver developers on AAPS, so I don't do much development in core AAPS. 
I wasn't even aware that there is a problem with Xdrip, but then I came accross it being mentioned in Release notes for upcoming release (the one that is being beta'd for last 2 or so months). 
It was mentioned few times on main (english) Facebook together with BYODA keyword (Build your own Dexcom App), it was also mentioned on discord (as BG being displayed with ?? instead of showing arrows). 
But there is no dedicated discussion about that, they are just pushing BYODA on us and tell everybody not to use Xdrip...

**Reply by savek-cc** [reply]
@andyrozman is this the reason why Milos states to migrate away from xDrip to BYODA? I was wondering what the technical reason for that was but couldn't find the actual issue anywhere...

**Reply by andyrozman** [reply]
@savek-cc There are 2 reasons:
1. Main one is that we get this weird data from Xdrip, that stops Loop for 1.5 days (and displays that red error triangle)
2. Secondary is that BYODA has better smoothing

Idea of getting away from Xdrip is putting a lot of people off, of installing the next version (when it comes out) and they started thinking (as well as me) on staying old version of AAPS

**Comment by Navid200** [comment]
> ... when new G6 sensor is started, old data (data accumulated before new sensor was started) is being added to NS ...
> 

I use G6 and I upload to Nightscout.
How can I recreate this problem and observe it?

**Reply by andyrozman** [reply]
You would need to install latest dev (beta) of AAPS and observe it when you change the sensor. We are not 100% sure if "duplicate" bad data comes from NS or from broadcast to AAPS.

**Comment by Navid200** [comment]
By the way, the issue https://github.com/NightscoutFoundation/xDrip/issues/1778 is related to starting a new transmitter, not starting a new sensor.

**Reply by andyrozman** [reply]
Yes I know, this is why I said its similar (it seems that Xdrip behaves similarly for both situations and when you start sensor/trasnmiter) it tries to "backfill" the data, and when solution for that is done (and if its applied for each sensor start) it should resolve the issue.

**Reply by Navid200** [reply]
Can I see this on Nightscout?
Can you explain how someone who is investigating can confirm the problem exists?

**Comment by Navid200** [comment]
If we have made a change to xDrip that has caused this problem, I need to know the xDrip release that has caused it.
This means you should be able to state that a version of xDrip is fine and does not cause this problem and the next release causes it.
I have never heard of anyone complaining about using xDrip with aaps.
So, if this is something new, I need to know what changed that caused it.

If there has been a change to aaps causing this problem, you should contact them and ask them to address the issue.

If there is a reason we need to change both xDrip and aaaps, I need a lot more than just a claim.
You can ask someone experienced to come here and explain what has changed in the code and why it is that the only way to address it is to change xDrip and not just aaps.

We cannot change xDrip just because someone asks for it.  So many different people use xDrip with so many different setups.  We need to be very careful with every single change we make.

**Reply by MilosKozak** [reply]
the issue was probably always there only AAPS was not doing so many data validations before

**Reply by MilosKozak** [reply]
i asked on discord to provide more info here

**Reply by Navid200** [reply]
Thanks

**Comment by stefanchavdarov** [comment]
Attaching logs from such event. I had 3 out of 3 sensor replacement behavior of like - 3.0 bg quality check plugin always flags it red with dup bg values. 
G6 with XDrip native being started - 2021-12-12 22:33:27
bug 1:
![image](https://user-images.githubusercontent.com/45970062/145732809-54475a0a-f7dd-425a-873e-5cf79cf16e76.png)
 - spam of same content lines at 22:28 help noone
bug 2:
I notice log several times during the 120 minute grace period telling "Agressively restarting collector service due to lack of reception". Its not lack of reception, but lack of BG readings and it shouldnt restart as it clearly isnt required.
bug 3:
at 2022-12-13 00:33 it misses the 1st reading for no reason (Could not read replay on auth message.)
![image](https://user-images.githubusercontent.com/45970062/145732830-e9066d42-821e-4c68-97e4-6720efad8306.png)

00:38:35.008 [pool-16-thread-4] D/BGSOURCE: [XdripPlugin$XdripWorker.doWork():80]: Received xDrip data: Bundle[{com.eveningoutpost.dexdrip.Extras.NoiseBlockLevel=200, 

Now the bug0:
![image](https://user-images.githubusercontent.com/45970062/145732845-d51c261b-16ca-45aa-af90-6421d64df5f3.png)
it would seem the last reading from old sendor (among other older) get duplicated, meaning it also sent historical readings along the new sensor 1st value

AAPS 3.0-beta9 logs:
[AndroidAPS_LOG_1639348900362.log.zip](https://github.com/Navid200/xDrip/files/7700032/AndroidAPS_LOG_1639348900362.log.zip)

I also sent the XDrip logs via app log upload option. I somehow didnt find a way to get the log file from phone

**Comment by MilosKozak** [comment]
what is visible in log:
```
22:23:34.638 [pool-16-thread-4] D/BGSOURCE: [XdripPlugin$XdripWorker.doWork():80]: Received xDrip data: Bundle[{com.eveningoutpost.dexdrip.Extras.NoiseBlockLevel=200, com.eveningoutpost.dexdrip.Extras.BgSlope=3.332233696213583E-6, com.eveningoutpost.dexdrip.Extras.VersionInfo=2111231541 9ff5104-2021.11.23, com.eveningoutpost.dexdrip.Extras.Raw=0.0, com.eveningoutpost.dexdrip.Extras.Time=1639340611328, com.eveningoutpost.dexdrip.Extras.CalibrationInfo=-1, com.eveningoutpost.dexdrip.Extras.NsNoiseLevel=null, com.eveningoutpost.dexdrip.Extras.Noise=-9999.0, com.eveningoutpost.dexdrip.Extras.SensorBattery=-1, com.eveningoutpost.dexdrip.Extras.NoiseWarning=0, com.eveningoutpost.dexdrip.Extras.BgSlopeName=Flat, com.eveningoutpost.dexdrip.Extras.BgEstimate=161.0, com.eveningoutpost.dexdrip.Extras.SourceDesc=G6 Native / G5 Native, com.eveningoutpost.dexdrip.Extras.SourceInfo=G6 Native, com.eveningoutpost.dexdrip.Extras.Collector.NanoStatus=  }]
22:23:34.649 [pool-16-thread-4] D/DATABASE: [XdripPlugin$XdripWorker.doWork():100]: Inserted bg GlucoseValue(id=11755, version=0, dateCreated=1639340614640, isValid=true, referenceId=null, interfaceIDs_backing=InterfaceIDs(nightscoutSystemId=null, nightscoutId=null, pumpType=null, pumpSerial=null, temporaryId=null, pumpId=null, startId=null, endId=null), timestamp=1639340611328, utcOffset=7200000, raw=0.0, value=161.0, trendArrow=FLAT, noise=null, sourceSensor=DEXCOM_G6_G5_NATIVE_XDRIP)
```

and 2 hours later:
```
00:38:40.855 [pool-16-thread-3] D/BGSOURCE: [XdripPlugin$XdripWorker.doWork():80]: Received xDrip data: Bundle[{com.eveningoutpost.dexdrip.Extras.NoiseBlockLevel=200, com.eveningoutpost.dexdrip.Extras.BgSlope=-8.205127153188827E-6, com.eveningoutpost.dexdrip.Extras.VersionInfo=2111231541 9ff5104-2021.11.23, com.eveningoutpost.dexdrip.Extras.Raw=0.0, com.eveningoutpost.dexdrip.Extras.Time=1639340611777, com.eveningoutpost.dexdrip.Extras.CalibrationInfo=-1, com.eveningoutpost.dexdrip.Extras.NsNoiseLevel=null, com.eveningoutpost.dexdrip.Extras.Noise=-9999.0, com.eveningoutpost.dexdrip.Extras.SensorBattery=-1, com.eveningoutpost.dexdrip.Extras.BgSlopeName=Flat, com.eveningoutpost.dexdrip.Extras.BgEstimate=161.0, com.eveningoutpost.dexdrip.Extras.SourceDesc=G6 Native / G5 Native, com.eveningoutpost.dexdrip.Extras.SourceInfo=G6 Native::Backfill, com.eveningoutpost.dexdrip.Extras.Collector.NanoStatus=  }]
00:38:40.870 [pool-16-thread-3] D/DATABASE: [XdripPlugin$XdripWorker.doWork():100]: Inserted bg GlucoseValue(id=11767, version=0, dateCreated=1639348720858, isValid=true, referenceId=null, interfaceIDs_backing=InterfaceIDs(nightscoutSystemId=null, nightscoutId=null, pumpType=null, pumpSerial=null, temporaryId=null, pumpId=null, startId=null, endId=null), timestamp=1639340611777, utcOffset=7200000, raw=0.0, value=161.0, trendArrow=FLAT, noise=null, sourceSensor=DEXCOM_G6_G5_NATIVE_XDRIP)
```

**Reply by stefanchavdarov** [reply]
@MilosKozak  Its worth saying the BG values validation plugin in AAPS is a good add. Multiple paths to transfer the data from xdrip to aaps resulting in duplicated bg readings to AAPS resulting in BG deviation of zero .... I experienced it 1st hand here once some time in the past and have seen few cases of likes from others.
.. This said, Dups either shouldnt happen (fighting xDrip front)... or/and should be filtered/ignored in AAPS if detected rather than failing the entire system. 

@Navid200 
On stop/start - What user is allowed to perform by app do and what that user should be doing are different topics. I suppose you are sick of telling users to wait and check for sensor stop to complete before starting new one so why not fit in app logic ... the right way

**Reply by Navid200** [reply]
Again, that's a different topic you can open a thread and we can see if there is a way to improve xDrip.

**Comment by MilosKozak** [comment]
there was not issue with xdrip before but the problem was there. just ignored by aaps. probably

**Reply by Navid200** [reply]
> there was not issue with xdrip before but the problem was there. just ignored by aaps. probably  

So, AAPS can continue to ignore it.
What burden was put on the shoulder of the patient by ignoring it then?

**Reply by MilosKozak** [reply]
well no. because delta (bg change) between about the same BGs is zero and this is passed to the algorithm

**Reply by MilosKozak** [reply]
zero deltas is the reason why we solve this issue

**Reply by Navid200** [reply]
I don't know what you are talking about it.  What delta is zero?

**Comment by Navid200** [comment]


**Reply by MilosKozak** [reply]
it should be resolved because this behaviour is wrong independent to AAPS

**Reply by MilosKozak** [reply]
idea: why to broadcast data from previous sensor at all?

**Reply by Navid200** [reply]
There are many behaviors of xDrip that may need to be modified.  If you list them all here, I will have no choice to delete them all.
I am running out of patience with this.
Again, please let's keep this thread on topic.

**Reply by Navid200** [reply]
> idea: why to broadcast data from previous sensor at all?

I don't know what you mean by that.
How can I recreate, and observe, this problem?

**Reply by tolot27** [reply]
> idea: why to broadcast data from previous sensor at all?

I have to dig into this problem. It would be best, if we can create a unit test for it. And indeed, I agree that we should fix this. The whole sync framework needs rework.

**Reply by Navid200** [reply]
@tolot27 Is there an open issue related to the work you have in mind?
Or, should I reopen this issue?
https://github.com/NightscoutFoundation/xDrip/issues/1890

**Reply by tolot27** [reply]
Yes, please reopen it.
